lib/hsa/mcwamp_hsa.cpp

```diff
@@ -2403,8 +2403,13 @@ HSADispatch::dispatchKernel(hsa_queue_t* commandQueue) {
     uint32_t queueMask = commandQueue->size - 1;
     // TODO: Need to check if package write is correct.
     uint64_t index = hsa_queue_load_write_index_relaxed(commandQueue);
+
+    uint64_t nextIndex = index + 1;
+    // spin-wait if the command queue is full
+    while(nextIndex - hsa_queue_load_read_index_acquire(commandQueue) >= commandQueue->size) ;
```

Comment:
Error on queue full is fine and an improv,went over current state of silently corrupting the queue.

On Jun 3, 2016, at 7:32 PM, Jack Chung <notifications@github.com<mailto:notifications@github.com>> wrote:


In lib/hsa/mcwamp_hsa.cpp<https://github.com/RadeonOpenCompute/hcc/pull/69#discussion_r65793182>:

> @@ -2403,8 +2403,13 @@ HSADispatch::dispatchKernel(hsa_queue_t* commandQueue) {
>      uint32_t queueMask = commandQueue->size - 1;
>      // TODO: Need to check if package write is correct.
>      uint64_t index = hsa_queue_load_write_index_relaxed(commandQueue);
> +
> +    uint64_t nextIndex = index + 1;
> +    // spin-wait if the command queue is full
> +    while(nextIndex - hsa_queue_load_read_index_acquire(commandQueue) >= commandQueue->size) ;


I'm feeling a bit uneasy here because it may potentially create an infinite loop in user mode without too much information disclosed to user applications.

Is it possible to fabricate a scenario where the queue is nearly full, or full? And measure how long the loop may take?

Also I'm wondering if there's a way (ex: signal / pipe / named pipe) we can interrupt the spin-wait and gracefully terminate the application?

-
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/RadeonOpenCompute/hcc/pull/69/files/4a163a4065807a2eb704b0b082113b8e053e215a#r65793182>, or mute the thread<https://github.com/notifications/unsubscribe/ACYSAss3bgFmaLXHi-i3_ZFIK-emMN_hks5qIMeAgaJpZM4ItDxr>.